core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -3104,23 +3104,106 @@ public static int position(ByteString seek, ByteString s) {
 
   /** SQL {@code POSITION(seek IN string FROM integer)} function. */
   public static int position(String seek, String s, int from) {
-    final int from0 = from - 1; // 0-based
-    if (from0 > s.length() || from0 < 0) {
+    if (from == 0) {
+      throw RESOURCE.fromNotZero().ex();
+    }
+    // Case when from is positive
+    if (from > 0) {
+      final int from0 = from - 1; // 0-based
+      if (from0 >= s.length() || from0 < 0) {
+        return 0;
+      } else {
+        return s.indexOf(seek, from0) + 1;
+      }
+    }
+    // Case when from is negative
+    final int rightIndex = from + s.length(); // negative position to positive index
```

Comment:
I don't think this is the correct semantics. BQ will actually search backward when `position` is negative. To see what I mean, try running `SELECT INSTR("fersher dude", "dude", -3, 1), INSTR("fersher dude", "dude", -4, 1), INSTR("fersher dude", "dude", -5, 1)` in http://pantheon/bigquery

Reply:
oooh boy great catch! fixing that now.

Reply:
Added some tests and a fix? It's definitely not particularly pretty though.